pkg/components/header.go

```diff
@@ -2,54 +2,99 @@ package components
-	return &HeaderComponent{}, nil
+// RequestHeader satisfies the NewComponent signature.
+func RequestHeader(_ context.Context, _ string, _ string, _ string) (interface{}, error) {
+	return &RequestHeaderComponent{}, nil
 }
 
 // Settings generates a config populated with defaults.
-func (*HeaderComponent) Settings() *HeaderConfig {
-	return &HeaderConfig{Names: []string{}, Values: []string{}}
+func (*RequestHeaderComponent) Settings() *RequestHeaderConfig {
+	return &RequestHeaderConfig{Headers: make(map[string][]string)}
 }
 
-func makeDecorator(name string, value string) transport.Decorator {
+func makeRequestDecorator(name string, value string) transport.Decorator {
 	return transport.NewHeader(func(*http.Request) (string, string) {
 		return name, value
 	})
 }
 
 // New generates the middleware.
-func (*HeaderComponent) New(_ context.Context, conf *HeaderConfig) (func(http.RoundTripper) http.RoundTripper, error) { // nolint
-	if len(conf.Names) != len(conf.Values) {
-		return nil, fmt.Errorf(
-			"header mismatch. %d names. %d values. these must match",
-			len(conf.Names),
-			len(conf.Values),
-		)
+func (*RequestHeaderComponent) New(_ context.Context, conf *RequestHeaderConfig) (func(http.RoundTripper) http.RoundTripper, error) { // nolint
+
+	length := 0
+	for _, headerValues := range conf.Headers {
+		length = length + len(headerValues)
+	}
+
+	ch := make(transport.Chain, 0, length)
+	for headerName, headerValues := range conf.Headers {
+		for _, headerValue := range headerValues {
+			ch = append(ch, makeRequestDecorator(headerName, headerValue))
+		}
 	}
-	ch := make(transport.Chain, 0, len(conf.Names))
-	for offset, name := range conf.Names {
-		ch = append(ch, makeDecorator(name, conf.Values[offset]))
+	return ch.Apply, nil
+}
+
+// ResponseHeaderConfig configures automated header injection.
+type ResponseHeaderConfig struct {
+	Headers map[string][]string `description:"Map of headers to inject into responses."`
+}
+
+// Name of the config root.
+func (*ResponseHeaderConfig) Name() string {
+	return "responseheaderinject"
+}
+
+// ResponseHeaderComponent implements the settings.Component interface.
+type ResponseHeaderComponent struct{}
+
+// ResponseHeader satisfies the NewComponent signature.
+func ResponseHeader(_ context.Context, a string, b string, c string) (interface{}, error) {
+	return &ResponseHeaderComponent{}, nil
+}
+
+// Settings generates a config populated with defaults.
+func (*ResponseHeaderComponent) Settings() *ResponseHeaderConfig {
+	return &ResponseHeaderConfig{Headers: make(map[string][]string)}
+}
+
+func makeResponseDecorator(name string, value string) transport.Decorator {
+	return transport.NewHeaders(nil, func(*http.Response) (string, string) {
+		return name, value
+	})
+}
+
+// New generates the middleware.
+func (*ResponseHeaderComponent) New(_ context.Context, conf *ResponseHeaderConfig) (func(http.RoundTripper) http.RoundTripper, error) { // nolint
```

Comment:
question: what linting constraint does this break? I noticed `//nolint` up above as well but this code this fine to me so I'm curious what linting rule this breaks.

Reply:
You have good eyes.  FWIW, I removed the `// nolint` comment and ran the linter.  No complaints.  So I guess it's a holdover from some prior lint violation.